File: lib/day-11/monkey-in-the-middle.clj
(ns day-11.monkey-in-the-middle
  (:require [clojure.string :as s]))

(def inp (slurp "lib/day-11/inp.txt"))
(def ex-inp (slurp "lib/day-11/ex-inp.txt"))

(defn to-int [i] (Integer/parseInt i))

(defn parse-starting-items [starting-items]
  (->> (s/split (last (s/split starting-items #": ")) #", ")
       (map to-int)
       (into [])))

(defn parse-operation [operation]
  (-> (str "(fn [old] ("
           (s/replace operation #"Operation: new = old " "")
           " old ))")
      (read-string)
      (eval)))

(defn parse-test [test]
  (let [tests (-> (apply str test) (s/replace #"[^0-9]+" ",") (s/split #","))
        [divisible-for m1 m2] (->> tests (drop 1) (map to-int))]
    {:test #(if (zero? (rem % divisible-for)) m1 m2)
     :divisible-for divisible-for}))

(defn parse-line [line]
  (let [[_ starting-items operation & test] (s/split-lines line)]
    (merge {:items (parse-starting-items starting-items)
            :operation (parse-operation operation)
            :inspected 0}
           (parse-test test))))

(defn parse-inp [text]
  (->> (s/split text #"\n\n")
       (map parse-line)
       (into [])))

(defn throw-item-to-other-monkeys [monkeys item monkey-num relief denominator]
  (let [monkey (get monkeys monkey-num)
        worry-lvl ((:operation monkey) item)
        worry-lvl (quot worry-lvl relief)
        worry-lvl (rem worry-lvl denominator)
        trow-to ((:test monkey) worry-lvl)]
    (update-in monkeys [trow-to :items] conj worry-lvl)))

(defn throw-all-item-to-other-monkeys [monkeys items monkey-num relief denominator]
  (->> items
       (reduce #(throw-item-to-other-monkeys %1 %2 monkey-num relief denominator) monkeys)))

(defn do-monkey-turn [monkeys monkey-num relief denominator]
  (let [items (get-in monkeys [monkey-num :items])]
    (-> monkeys
        (throw-all-item-to-other-monkeys items monkey-num relief denominator)
        (update-in [monkey-num :items] (constantly []))
        (update-in [monkey-num :inspected] + (count items)))))

(defn do-turn [monkeys _curr_turn relief denominator]
  (->> (count monkeys)
       (range)
       (reduce #(do-monkey-turn %1 %2 relief denominator) monkeys)))

(defn do-turns [text turns relief]
  (let [monkeys (parse-inp text)
        denominator (->> monkeys (map :divisible-for) (reduce *))]
    (->> (range turns)
         (reduce #(do-turn %1 %2 relief denominator) monkeys)
         (map :inspected)
         (sort >)
         (take 2)
         (apply *))))

(defn ex1 [text] (do-turns text 20 3))
(defn ex2 [text] (do-turns text 10000 1))

(ex1 ex-inp) ;; 10605
(ex2 ex-inp) ;; 2713310158
(ex1 inp)    ;; 55458
(ex2 inp)    ;; 14508081294

File: elixir/lib/day-11/monkey_in_the_middle.ex
defmodule MonkeyInTheMiddle do
  def parse_input(text) do
    text
    |> String.split("\n\n")
    |> Enum.map(&parse_block/1)
    |> Enum.into(%{})
  end

  defp parse_block(block) do
    block
    |> String.split("\n")
    |> Enum.reduce({nil, %{throws: 0}}, &parse_block_line/2)
  end

  defp parse_block_line(line, {idx, acc}) do
    case line do
      "Monkey " <> rest ->
        {idx, _} = rest |> Integer.parse()
        {idx, acc}

      "  Starting items: " <> rest ->
        starting_items = rest |> String.split(", ") |> Enum.map(&String.to_integer/1)
        {idx, Map.put(acc, :items, starting_items)}

      "  Operation: new =" <> rest ->
        throw_to = Code.eval_string("fn old -> " <> rest <> " end") |> elem(0)
        {idx, Map.put(acc, :operation, throw_to)}

      "  Test: divisible by " <> rest ->
        {idx, Map.put(acc, :div, String.to_integer(rest))}

      "    If true: throw to monkey " <> rest ->
        {idx, Map.put(acc, true, String.to_integer(rest))}

      "    If false: throw to monkey " <> rest ->
        {idx, Map.put(acc, false, String.to_integer(rest))}

      _ ->
        {idx, acc}
    end
  end

  def monkeys_turn(monkeys, rounds, calm_with) do
    Enum.reduce(1..rounds, monkeys, fn _round, monkeys ->
      Enum.reduce(monkeys, monkeys, fn {trowing_monkey, _old_monkey}, monkeys ->
        monkey = monkeys[trowing_monkey]

        monkey.items
        |> Enum.reduce(monkeys, fn item, monkeys ->
          worry_lvl = monkey.operation.(item)
          worry_lvl_calmed = calm_with.(worry_lvl)
          divisible? = rem(worry_lvl_calmed, monkey.div) == 0
          throw_to = monkey[divisible?]

          update_in(monkeys, [throw_to, :items], &[worry_lvl_calmed | &1])
        end)
        |> update_in([trowing_monkey, :throws], &(&1 + length(monkey.items)))
        |> put_in([trowing_monkey, :items], [])
      end)
    end)
    |> Enum.map(fn {_k, v} -> v.throws end)
    |> Enum.sort(&>/2)
    |> Enum.take(2)
    |> Enum.reduce(1, &*/2)
  end

  def ex1(monkeys) do
    monkeys_turn(monkeys, 20, &div(&1,3))
  end

  def ex2(monkeys) do
    scd =
      monkeys
      |> Enum.map(fn {_, %{div: div}} -> div end)
      |> Enum.reduce(1, &*/2)

    monkeys_turn(monkeys, 10000, &rem(&1,scd))
  end
end

ex_inp = File.read!("./lib/day-11/ex-inp.txt")
inp = File.read!("./lib/day-11/inp.txt")



# 10605
ex_inp
|> MonkeyInTheMiddle.parse_input()
|> MonkeyInTheMiddle.ex1()
|> IO.inspect()

# 55458
inp
|> MonkeyInTheMiddle.parse_input()
|> MonkeyInTheMiddle.ex1()
|> IO.inspect()

# 2713310158
ex_inp
|> MonkeyInTheMiddle.parse_input()
|> MonkeyInTheMiddle.ex2()
|> IO.inspect()

# 14508081294
inp
|> MonkeyInTheMiddle.parse_input()
|> MonkeyInTheMiddle.ex2()
|> IO.inspect()
